Initial Linux nightly bundles upload

SomeoneToIgnore opened the pull request:
Changes Zed CI to build and upload Linux nightly bundles.

* `todo!(linux)` are replaced with `TODO linux` to make `todo!`-based workflows more convenient
* renames `run-build-dmg` label into `run-bundling`, also renames a few GH Actions entries to be more generic
* split the bundle uploading script into two phases, so that first all bundles are built and uploaded, and only after all such uploads are fine, their nightly version SHA is bumped.
The approach is not ideal: if part of the packages uploads and part fails, we would not promote new versions, but neither will we revert the uploaded ones, so somebody might get newer versions under the old SHA, if queries the server for downloads.

* adds a `*.deb` package building with a couple of caveats, marked with new `TODO linux` entries:

1. `cargo-bundle` is not very flexible, so it generates artifacts with the structure and names that we're unable to alter before/during the generation.
For that, a set of extra steps is made by repacking the *.deb package — this is not very portable between different Linux distros, so later one needs to find a way to combine multiple package types in this script.

2. `cargo-bundle` is not able to properly generate the *.msi bundle despite declaring it in the features: https://github.com/burtonageo/cargo-bundle/issues/116
Windows needs to invent its own way of bundling or fix the tool.

3. Both `cli` and `zed` binaries are added into the archive under `/usr/local/bin/` path with their `-$channel` suffix (-nightly/-preview/-dev/-stable) and a `/usr/local/bin/zed -> /usr/local/bin/cli-nightly` symlink is made to make CLI work as Zed launcher:
```
~/work/zed kb/linux-nightly:origin/kb/linux-nightly*​ ❯ dpkg -c target/zed_amd64.deb 
drwxr-xr-x allaptop/allaptop 0 2024-03-06 00:53 ./
drwxr-xr-x allaptop/allaptop 0 2024-03-06 00:53 ./usr/
drwxr-xr-x allaptop/allaptop 0 2024-03-06 00:53 ./usr/local/
drwxr-xr-x allaptop/allaptop 0 2024-03-06 00:53 ./usr/local/bin/
-rwxr-xr-x allaptop/allaptop 8746832 2024-03-06 00:53 ./usr/local/bin/cli-nightly
-rwxr-xr-x allaptop/allaptop 689078560 2024-03-06 00:53 ./usr/local/bin/zed-nightly
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/applications/
-rw-r--r-- allaptop/allaptop       153 2024-03-06 00:53 ./usr/share/applications/zed.desktop
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/icons/
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/icons/hicolor/
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/icons/hicolor/1024x1024@2x/
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/icons/hicolor/1024x1024@2x/apps/
-rw-r--r-- allaptop/allaptop    716288 2024-03-06 00:53 ./usr/share/icons/hicolor/1024x1024@2x/apps/zed.png
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/icons/hicolor/512x512/
drwxr-xr-x allaptop/allaptop         0 2024-03-06 00:53 ./usr/share/icons/hicolor/512x512/apps/
-rw-r--r-- allaptop/allaptop    239870 2024-03-06 00:53 ./usr/share/icons/hicolor/512x512/apps/zed.png
lrwxrwxrwx allaptop/allaptop         0 2024-03-06 00:53 ./usr/local/bin/zed -> /usr/local/bin/cli-nightly
```

But the CLI does not work under Linux yet and there's no way to install that CLI from Zed now; Zed binary itself is not able to open `file/location:12:34`-like things and set up the env properly, but is able to start or open a directory.

So, this structure can be considered temporary and changed, if needed.

4. Zed Nightly on Linux does not know how to update itself, so all nightly publishing is not picked up automatically.

5. Rust cache from `main` builds does not get shared between CI jobs, due to being run in a different CI job that forms a different CI key, so 
```
      - name: Cache dependencies
        uses: swatinem/rust-cache@v2
        with:
          save-if: ${{ false }}
```
would not work.
This makes Linux bundling jobs long.

Release Notes:

- N/A

SomeoneToIgnore committed:
Use more generic names for bundling items

SomeoneToIgnore committed:
Draft *.deb bundling of Zed binary

SomeoneToIgnore committed:
Fill `authors` manifest field in `zed`

cargo-bundle will reuse this field in the `Maintainers` debian package metadata.

SomeoneToIgnore committed:
Repackage Zed archive properly

SomeoneToIgnore committed:
Work on *.deb uploads

SomeoneToIgnore committed:
Split bundle uploading and promotion into separate steps

SomeoneToIgnore committed:
Add regular (`main` and label-based) Linux bundling capabilities

SomeoneToIgnore committed:
Install linux dependencies

SomeoneToIgnore committed:
Bump upload action versions

SomeoneToIgnore committed:
Try to use the Rust cache for Linux bundling jobs

SomeoneToIgnore committed:
Use the same dir for final *.deb as *.dmg bundling does

kvark commented on .github/workflows/ci.yml:
```diff
@@ -232,3 +232,81 @@ jobs:
+
+  bundle-deb:
+    name: Create a *.deb Linux bundle
+    runs-on: ubuntu-latest
```
If we build it on latest ubuntu, wouldn't this prevent anyone on an older linux to run it? I.e. it will expect a certain high version of GLIBC to be available, even though we unlikely actually need it. Perhaps, we should use something like `ubuntu-18` here?

SomeoneToIgnore replied:
I think we need much more in the end, here's some good example: https://github.com/rust-lang/rust-analyzer/blob/767d5d3eab237a1fbc10d62b120020b12f6026c5/.github/workflows/release.yaml#L26-L51

We can already add 

```
          - os: ubuntu-20.04
            target: x86_64-unknown-linux-gnu
            code-target: linux-x64
            container: ubuntu:18.04
          - os: ubuntu-20.04
            target: aarch64-unknown-linux-gnu
            code-target: linux-arm64
          - os: ubuntu-20.04
            target: arm-unknown-linux-gnueabihf
            code-target: linux-armhf
```

and whatever ubuntu-18 libc, but this does not seem hard for the community to do, if they have a need for the particular one (maybe, it also needs libc added into the name?).

What's harder is the caching story, so I would rather concentrate on that myself (probably, in standalone PRs).

SomeoneToIgnore replied:
But reading this again, we need to set the base version low enough to get builds more universal?
Sure, let's try that.

kvark replied:
yep, just this

SomeoneToIgnore replied:
Apparently, it's gone from the runners: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners

<img width="839" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/a657fd0c-483b-4c98-a8ac-7be4e7e35f98">

The job does not start: https://github.com/zed-industries/zed/actions/runs/8167955494/job/22329289230?pr=8913

But I can use the 20.04 then.

SomeoneToIgnore committed:
Use older Ubuntu when bundling for Linux

This way, older libc versions are used, which are more compatible with various Linux distros.

burrbull commented:
```
zed: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
```

SomeoneToIgnore commented:
```
ldd /usr/local/bin/zed
	linux-vdso.so.1 (0x00007ffd73385000)
	libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007f0acc3ef000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007f0acc000000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f0acc3d5000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f0acc3b1000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f0acc3ac000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f0acbe1e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f0acc57d000)
```

It sounds odd, but all seems to work on my Fedora (dpkg is installable there).
Could somebody provide more details about the system and libssl installation path, if it breaks; or confirm that it works?

We could also consider static linking of this library, but I'm out of platform dev context to know if it'll work.

burrbull commented:
> ```
> zed: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
> ```

Ubuntu 23.10
Previous build (before CI version decrease) works fine.

SomeoneToIgnore commented:
Here's what I have installed in the system

```
> ls -lah /lib64/libssl*
-rwxr-xr-x. 1 root root 421K Dec 21 02:00 /lib64/libssl3.so
lrwxrwxrwx. 1 root root   15 Aug 31  2023 /lib64/libssl.so -> libssl.so.3.1.1
lrwxrwxrwx. 1 root root   16 Jul 20  2023 /lib64/libssl.so.1.1 -> libssl.so.1.1.1q
-rwxr-xr-x. 1 root root 630K Jul 20  2023 /lib64/libssl.so.1.1.1q
lrwxrwxrwx. 1 root root   15 Aug 31  2023 /lib64/libssl.so.3 -> libssl.so.3.1.1
-rwxr-xr-x. 1 root root 661K Aug 31  2023 /lib64/libssl.so.3.1.1

> sudo dnf repoquery --userinstalled|rg -i ssl
openssl-1:3.1.1-4.fc39.x86_64
openssl-devel-1:3.1.1-4.fc39.x86_64
```

And here's the freshmost *.deb from previous CI builds:
```
ldd /usr/local/bin/zed
	linux-vdso.so.1 (0x00007fff5eb6f000)
	libssl.so.3 => /lib64/libssl.so.3 (0x00007f87b788e000)
	libcrypto.so.3 => /lib64/libcrypto.so.3 (0x00007f87b7200000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f87b7874000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f87b7850000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f87b701e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f87b7a25000)
```

So the difference seems to be related to libssl version the binary is linked with.
I expect, the next launch error after we fix that is `libcrypto.so.3`

I will try to bump the version on the CI with old ubuntu and make it work, but overall the dynamic dependencies' situation in Linux seems sad to catch up with, so whatever I'll pick, there will be unhappy users.
I wonder what makes those non-statically linked in our case and why do we do that.

SomeoneToIgnore committed:
Rollback to Ubuntu 22.04 to keep up with the modern users.

Ubuntu 20.04 uses older libcrypto and libssl which we link with dynamically, which Ubuntu 22.04 does not have by default, causing runtime issues.

SomeoneToIgnore committed:
Split mac and linux nightly uploads entirely

SomeoneToIgnore commented:
Ok, I'm back to Ubuntu 22.04 as more minestream as an example solution, but the point of the whole PR is to make a template and encourage better platform-knowledgeable people to influence how those builds are delivered to their platforms.

So far, we know that per a Zed Linux release, users need different libc + N dynamicly linked libraries (at least 2 crypto ones) + package types (rpm, deb, flatpack, source archive with submodules, etc.) in many combinations.

From Zed side, we need to deliver Zed editor binary, a bunch or resources (icons at least, more files for macOs) and, optionally, cli binary (as it's not clear whether that's going to be used outside macOs).

Next step (concurrent) would be to:
* gather better detailed requirements on which delivery artifacts (and related workflows) are of most priority
* decide how to do autoupdates for Linux and, preferrably, Windows too.
* consider how cli is supposed to work (if supposed) on Linux and, preferrably, Windows too.

and move on based on that, accepting normal relevant PRs on the topic.
